impl/spanner/batch/batch_test.go

```diff
@@ -0,0 +1,160 @@
+		t.Fatal(err)
+	}
+	client, cleanup := testutil.CreateDatabase(ctx, t, ddl)
+	b := New(client)
+
+	if err := directory.New(client).Write(ctx, &dtype.Directory{
+		DirectoryID: dirID,
+		Map:         &tpb.Tree{},
+		Log:         &tpb.Tree{},
+		VRFPriv:     &keyspb.PrivateKey{Der: []byte("privkeybytes")},
+	}); err != nil {
+		t.Fatalf("directories.Write(%v): %v", dirID, err)
+	}
+
+	return b, cleanup
+}
+
+func TestNewForTest(t *testing.T) {
+	ctx := context.Background()
+	directoryID := "new"
+	_, done := NewForTest(ctx, t, directoryID)
+	defer done()
+}
+
+func TestBatchIntegration(t *testing.T) {
+	storageFactory :=
+		func(ctx context.Context, t *testing.T, dirID string) (sequencer.Batcher, func(context.Context)) {
+			b, done := NewForTest(ctx, t, dirID)
+			return b, func(_ context.Context) { done() }
+		}
+
+	storagetest.RunBatchStorageTests(t, storageFactory)
+}
+
+func TestWriteBatch(t *testing.T) {
+	ctx := context.Background()
+	directoryID := "writebatch"
+	m, done := NewForTest(ctx, t, directoryID)
+	defer done()
+	for _, tc := range []struct {
+		rev     int64
+		wantErr bool
+		sources []*spb.MapMetadata_SourceSlice
+	}{
+		// Tests are cumulative.
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 10}}},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 11}}, wantErr: true},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{{LogId: 2, HighestExclusive: 20}}, wantErr: true},
+		{rev: 0, sources: []*spb.MapMetadata_SourceSlice{}, wantErr: true},
+		{rev: 1, sources: []*spb.MapMetadata_SourceSlice{}},
+		{rev: 1, sources: []*spb.MapMetadata_SourceSlice{{LogId: 1, HighestExclusive: 10}}, wantErr: true},
+	} {
+		err := m.WriteBatchSources(ctx, directoryID, tc.rev, &spb.MapMetadata{Sources: tc.sources})
+		if got, want := err != nil, tc.wantErr; got != want {
+			t.Errorf("WriteBatchSources(%v, %v): err: %v. code: %v, want %v",
+				tc.rev, tc.sources, err, got, want)
+		}
+	}
+}
+
+func TestReadBatch(t *testing.T) {
+	ctx := context.Background()
+	directoryID := "readbatch"
+	m, done := NewForTest(ctx, t, directoryID)
+	defer done()
+
+	for _, tc := range []struct {
+		rev  int64
+		want *spb.MapMetadata
+	}{
+		{rev: 0, want: &spb.MapMetadata{Sources: []*spb.MapMetadata_SourceSlice{
+			{LogId: 1, HighestExclusive: 10},
+			{LogId: 2, HighestExclusive: 20},
+		}}},
+		{rev: 1, want: &spb.MapMetadata{Sources: []*spb.MapMetadata_SourceSlice{
+			{LogId: 1, HighestExclusive: 11},
+			{LogId: 2, HighestExclusive: 22},
+		}}},
+	} {
+		if err := m.WriteBatchSources(ctx, directoryID, tc.rev, tc.want); err != nil {
+			t.Fatalf("WriteBatch(%v): %v", tc.rev, err)
+		}
+		got, err := m.ReadBatch(ctx, directoryID, tc.rev)
+		if err != nil {
+			t.Fatalf("ReadBatch(%v): %v", tc.rev, err)
+		}
+		if !cmp.Equal(got, tc.want, cmp.Comparer(proto.Equal)) {
+			t.Errorf("ReadBatch(%v): %v, want %v", tc.rev, got, tc.want)
+		}
+	}
+}
+
+func TestHightestRev(t *testing.T) {
```

Comment:
sp: s/Hightest/Highest/ unless pun is intentional. In which case, well played.

Reply:
Thanks for the catch